daemon/container.go

```diff
@@ -1052,7 +1052,7 @@ func (container *Container) setupContainerDns() error {
 		return err
 	}
 
-	if config.NetworkMode != "host" {
+	if !config.NetworkMode.IsHost() || !config.NetworkMode.IsContainer() {
```

Comment:
And maybe instead of even that, we can do an `IsShared()`, assuming anything that is sharing can't change existing network options, and get rid of the need for multiple checks in `runconfig.Parse`